Add backup plugins (#5985)
kyma-project/kyma: components/backup-plugins/README.md

# Backup plugins

## Overview

> Backup plugins provide an implementation of a few plugins required to properly restore the Kyma cluster and resources created inside it using Velero. They focus mainly on resources related to the Service Catalog, such as instances or bindings. Each plugin is defined in a separate file inside the [`internal/plugins`](internal/plugins) folder. The purpose of each plugin is defined in the file's comments.

Reviewer: ```suggestion
Backup plugins provide the functionality necessary to properly restore the Kyma cluster and its resources using Velero. They focus mainly on resources related to the Service Catalog, such as instances or bindings. Each plugin is defined in a separate file inside the [`internal/plugins`](internal/plugins) folder. 
```
I think the last sentence is not necessary.

The structure of folders and files is based on the [Velero plugin example repository](https://github.com/heptio/velero-plugin-example).

## Installation

> Run a Docker image build from the [Dockerfile](Dockerfile) as an init container inside the deployment of the [Velero chart](../../resources/velero). During development, you can push a built image to your own image repository and run:

Reviewer: ```suggestion
Run a Docker image build from the [Dockerfile](Dockerfile) as an init container inside the [Velero chart](../../resources/velero) Deployment. 
During development, you can push the built image to your own image repository and run:
```

```bash
velero plugin add {yourRepo/imageName:tag}
```

## Create a new plugin

> - New plugins must be added under [`internal/plugins`](internal/plugins) :

Reviewer: ```suggestion
To add a new plugin, perform the following steps:
1. Go to the [`internal/plugins`](internal/plugins) folder: 
```

```             
  ├── internal                                                                  
    ├── plugins
      ├── backup    # new Backup Item Action plugins 
      ├── restore   # new Restore Item Action plugins 
  ```

> - All new plugins must be registered in the `components/backup-plugins/main.go` file.

Reviewer: ```suggestion
3. Register created plugins in the `components/backup-plugins/main.go` file.
```
I would make this the third point.

- **Backup Item Action** - performs arbitrary logic on individual items prior to storing them in the backup file.

> - **Restore Item Action** - performs arbitrary logic on individual items prior to restoring them in the Kyma cluster.

Reviewer: I'm wondering if we should leave these two here or move them to the overview section. They can be left here, but then maybe we need a kind of introduction, such as:

This folder includes the `backup` and `restore` subfolders, where you can define the following plugins:

Question: is the `Backup Item Action/ Restore Item Action` an object?

> - All plugins must be implemented as the following example:

Reviewer: ```suggestion
2. Implement plugins as in the following example:
```

# Backup plugins

## Overview

Backup plugins provide the functionality necessary to properly restore the Kyma cluster and its resources using Velero. They focus mainly on resources related to the Service Catalog, such as instances or bindings. Each plugin is defined in a separate file inside the [`internal/plugins`](internal/plugins) folder.

The structure of folders and files is based on the [Velero plugin example repository](https://github.com/heptio/velero-plugin-example).

## Installation

Run a Docker image build from the [Dockerfile](Dockerfile) as an init container inside the [Velero chart](../../resources/velero) Deployment.

> During development, you can push the built image to your own image repository and run:

Reviewer: ```suggestion
During development, you can push the image to your own image repository and add the plugin by running:
```

```bash
velero plugin add {yourRepo/imageName:tag}
```

## Development

### Create a new plugin

> To add a new plugin, perform the following steps:

Reviewer: ```suggestion
To create a new plugin for Velero to use, perform the following steps:
```

1. Go to the [`internal/plugins`](internal/plugins) folder:

> This folder includes the backup and restore subfolders, where you can define the following plugins:

Reviewer: ```suggestion
This folder includes the backup and restore subfolders, where you can define plugins based on the following object types:
```

> - **Backup Item Action** - performs arbitrary logic on individual items prior to storing them in the backup file.

Reviewer: ```suggestion
- **Backup Item Action** - performs arbitrary logic on individual items before storing them in the backup file.
```

> - **Restore Item Action** - performs arbitrary logic on individual items prior to restoring them in the Kyma cluster.

Reviewer: ```suggestion
- **Restore Item Action** - performs arbitrary logic on individual items before restoring them in the Kyma cluster.
```
